---

_You are currently looking at **version 1.1** of this notebook. To download notebooks and datafiles, as well as get help on Jupyter notebooks in the Coursera platform, visit the [Jupyter Notebook FAQ](https://www.coursera.org/learn/python-data-analysis/resources/0dhYG) course resource._

---

In [1]:
import pandas as pd
import numpy as np
from scipy.stats import ttest_ind

# Assignment 4 - Hypothesis Testing
This assignment requires more individual learning than previous assignments - you are encouraged to check out the [pandas documentation](http://pandas.pydata.org/pandas-docs/stable/) to find functions or methods you might not have used yet, or ask questions on [Stack Overflow](http://stackoverflow.com/) and tag them as pandas and python related. And of course, the discussion forums are open for interaction with your peers and the course staff.

Definitions:
* A _quarter_ is a specific three month period, Q1 is January through March, Q2 is April through June, Q3 is July through September, Q4 is October through December.
* A _recession_ is defined as starting with two consecutive quarters of GDP decline, and ending with two consecutive quarters of GDP growth.
* A _recession bottom_ is the quarter within a recession which had the lowest GDP.
* A _university town_ is a city which has a high percentage of university students compared to the total population of the city.

**Hypothesis**: University towns have their mean housing prices less effected by recessions. Run a t-test to compare the ratio of the mean price of houses in university towns the quarter before the recession starts compared to the recession bottom. (`price_ratio=quarter_before_recession/recession_bottom`)

The following data files are available for this assignment:
* From the [Zillow research data site](http://www.zillow.com/research/data/) there is housing data for the United States. In particular the datafile for [all homes at a city level](http://files.zillowstatic.com/research/public/City/City_Zhvi_AllHomes.csv), ```City_Zhvi_AllHomes.csv```, has median home sale prices at a fine grained level.
* From the Wikipedia page on college towns is a list of [university towns in the United States](https://en.wikipedia.org/wiki/List_of_college_towns#College_towns_in_the_United_States) which has been copy and pasted into the file ```university_towns.txt```.
* From Bureau of Economic Analysis, US Department of Commerce, the [GDP over time](http://www.bea.gov/national/index.htm#gdp) of the United States in current dollars (use the chained value in 2009 dollars), in quarterly intervals, in the file ```gdplev.xls```. For this assignment, only look at GDP data from the first quarter of 2000 onward.

Each function in this assignment below is worth 10%, with the exception of ```run_ttest()```, which is worth 50%.

In [3]:
# Content of .txt file
!cat university_towns.txt

Alabama[edit]
Auburn (Auburn University)[1]
Florence (University of North Alabama)
Jacksonville (Jacksonville State University)[2]
Livingston (University of West Alabama)[2]
Montevallo (University of Montevallo)[2]
Troy (Troy University)[2]
Tuscaloosa (University of Alabama, Stillman College, Shelton State)[3][4]
Tuskegee (Tuskegee University)[5]
Alaska[edit]
Fairbanks (University of Alaska Fairbanks)[2]
Arizona[edit]
Flagstaff (Northern Arizona University)[6]
Tempe (Arizona State University)
Tucson (University of Arizona)
Arkansas[edit]
Arkadelphia (Henderson State University, Ouachita Baptist University)[2]
Conway (Central Baptist College, Hendrix College, University of Central Arkansas)[2]
Fayetteville (University of Arkansas)[7]
Jonesboro (Arkansas State University)[8]
Magnolia (Southern Arkansas University)[2]
Monticello (University of Arkansas at Monticello)[2]
Russellville (Arkansas Tech University)[2]
Searcy (Harding University)[5]
California[edit]
Angwin (Pacific Union College

In [129]:
# Reading txt file into data frame
df_uni = pd.read_table('university_towns.txt', header=None)
df_uni.columns = ['text']
df_uni.shape

(567, 1)

In [130]:
# Strip state from cell
df_uni['State'] = df_uni['text'].apply(lambda x: x.split('[')[0].strip() \
              if x.count('[edit]') > 0 else np.NaN).fillna(method="ffill")

# Strip region from cell
df_uni['RegionName'] = df_uni['text'].apply(lambda x: x.split('(')[0] \
                            .strip() if x.count('(') >= 0 else np.NaN)    
    
df_uni['RegionName'] = df_uni['RegionName'].apply(lambda x: np.NaN \
                                     if x.count('[edit]') > 0 else x)

# Drop initial text column and reset index
df_uni = df_uni.dropna().drop('text', axis=1) \
            .reindex(columns=['State', 'RegionName']).reset_index(drop=True)

In [131]:
df_uni.head()

State    RegionName
0  Alabama        Auburn
1  Alabama      Florence
2  Alabama  Jacksonville
3  Alabama    Livingston
4  Alabama    Montevallo

In [105]:
# Use this dictionary to map state names to two letter acronyms
states = {'OH': 'Ohio', 'KY': 'Kentucky', 'AS': 'American Samoa', 'NV': 'Nevada', 'WY': 'Wyoming', 'NA': 'National', 'AL': 'Alabama', 'MD': 'Maryland', 'AK': 'Alaska', 'UT': 'Utah', 'OR': 'Oregon', 'MT': 'Montana', 'IL': 'Illinois', 'TN': 'Tennessee', 'DC': 'District of Columbia', 'VT': 'Vermont', 'ID': 'Idaho', 'AR': 'Arkansas', 'ME': 'Maine', 'WA': 'Washington', 'HI': 'Hawaii', 'WI': 'Wisconsin', 'MI': 'Michigan', 'IN': 'Indiana', 'NJ': 'New Jersey', 'AZ': 'Arizona', 'GU': 'Guam', 'MS': 'Mississippi', 'PR': 'Puerto Rico', 'NC': 'North Carolina', 'TX': 'Texas', 'SD': 'South Dakota', 'MP': 'Northern Mariana Islands', 'IA': 'Iowa', 'MO': 'Missouri', 'CT': 'Connecticut', 'WV': 'West Virginia', 'SC': 'South Carolina', 'LA': 'Louisiana', 'KS': 'Kansas', 'NY': 'New York', 'NE': 'Nebraska', 'OK': 'Oklahoma', 'FL': 'Florida', 'CA': 'California', 'CO': 'Colorado', 'PA': 'Pennsylvania', 'DE': 'Delaware', 'NM': 'New Mexico', 'RI': 'Rhode Island', 'MN': 'Minnesota', 'VI': 'Virgin Islands', 'NH': 'New Hampshire', 'MA': 'Massachusetts', 'GA': 'Georgia', 'ND': 'North Dakota', 'VA': 'Virginia'}

# Map inverse dictionary value: key
states_inv = {v: k for k, v in states.items()}

# Map states to acronyms -> Not necessary
#df_uni['state_XX'] = df_uni['State'].replace(states_inv)
#df_uni.tail()

State     RegionName state_XX
512  Wisconsin    River Falls       WI
513  Wisconsin  Stevens Point       WI
514  Wisconsin       Waukesha       WI
515  Wisconsin     Whitewater       WI
516    Wyoming        Laramie       WY

In [8]:
def get_list_of_university_towns():
    '''Returns a DataFrame of towns and the states they are in from the 
    university_towns.txt list. The format of the DataFrame should be:
    DataFrame( [ ["Michigan", "Ann Arbor"], ["Michigan", "Yipsilanti"] ], 
    columns=["State", "RegionName"]  )
    
    The following cleaning needs to be done:

    1. For "State", removing characters from "[" to the end.
    2. For "RegionName", when applicable, removing every character from 
    " (" to the end.
    3. Depending on how you read the data, you may need to remove 
    newline character '\n'. '''
    
    # Reading txt file into data frame
    df_uni = pd.read_table('university_towns.txt', header=None)
    df_uni.columns = ['text']
    df_uni.head()
    
    # Strip state from cell
    df_uni['State'] = df_uni['text'].apply(lambda x: x.split('[')[0].strip() \
                if x.count('[edit]') > 0 else np.NaN).fillna(method="ffill")

    # Strip region from cell (aware with the x.count('(') >= 0 )
    df_uni['RegionName'] = df_uni['text'].apply(lambda x: x.split('(')[0] \
                               .strip() if x.count('(') >= 0 else np.NaN)    

    df_uni['RegionName'] = df_uni['RegionName'].apply(lambda x: np.NaN \
                                     if x.count('[edit]') > 0 else x)

    # Drop initial text column and reset index
    df_uni = df_uni.dropna().drop('text', axis=1) \
            .reindex(columns=['State', 'RegionName']).reset_index(drop=True)
    
    return df_uni

In [9]:
get_list_of_university_towns().shape

(517, 2)

#Convert to code to debug
import re
import pandas as pd
import numpy as np
# list of unique states
stateStr = """
Ohio, Kentucky, American Samoa, Nevada, Wyoming
,National, Alabama, Maryland, Alaska, Utah
,Oregon, Montana, Illinois, Tennessee, District of Columbia
,Vermont, Idaho, Arkansas, Maine, Washington
,Hawaii, Wisconsin, Michigan, Indiana, New Jersey
,Arizona, Guam, Mississippi, Puerto Rico, North Carolina
,Texas, South Dakota, Northern Mariana Islands, Iowa, Missouri
,Connecticut, West Virginia, South Carolina, Louisiana, Kansas
,New York, Nebraska, Oklahoma, Florida, California
,Colorado, Pennsylvania, Delaware, New Mexico, Rhode Island
,Minnesota, Virgin Islands, New Hampshire, Massachusetts, Georgia
,North Dakota, Virginia
"""
#list of regionName entries string length
regNmLenStr = """
06,08,12,10,10,04,10,08,09,09,05,06,11,06,12,09,08,10,12,06,
06,06,08,05,09,06,05,06,10,28,06,06,09,06,08,09,10,35,09,15,
13,10,07,21,08,07,07,07,12,06,14,07,08,16,09,10,11,09,10,06,
11,05,06,09,10,12,06,06,11,07,08,13,07,11,05,06,06,07,10,08,
11,08,13,12,06,04,08,10,08,07,12,05,06,09,07,10,16,10,06,12,
08,07,06,06,06,11,14,11,07,06,06,12,08,10,11,06,10,14,04,11,
18,07,07,08,09,06,13,11,12,10,07,12,07,04,08,09,09,13,08,10,
16,09,10,08,06,08,12,07,11,09,07,09,06,12,06,09,07,10,09,10,
09,06,15,05,10,09,11,12,10,10,09,13,06,09,11,06,11,09,13,37,
06,13,06,09,49,07,11,12,09,11,11,07,12,10,06,06,09,04,09,15,
10,12,05,09,08,09,09,07,14,06,07,16,12,09,07,09,06,32,07,08,
08,06,10,36,09,10,09,06,09,11,09,06,10,07,14,08,07,06,10,09,
05,11,07,06,08,07,05,07,07,04,06,05,09,04,25,06,07,08,05,08,
06,05,11,09,07,07,06,13,09,05,16,05,10,09,08,11,06,06,06,10,
09,07,06,07,10,05,08,07,06,08,06,30,09,07,06,11,07,12,08,09,
16,12,11,08,06,04,10,10,15,05,11,11,09,08,06,04,10,10,07,09,
11,08,26,07,13,05,11,03,08,07,06,05,08,13,10,08,08,08,07,07,
09,05,04,11,11,07,06,10,11,03,04,06,06,08,08,06,10,09,05,11,
07,09,06,12,13,09,10,11,08,07,07,08,09,10,08,10,08,56,07,12,
07,16,08,04,10,10,10,10,07,09,08,09,09,10,07,09,09,09,12,14,
10,29,19,07,11,12,13,13,09,10,12,12,12,08,10,07,10,07,07,08,
08,08,09,10,09,11,09,07,09,10,11,11,10,09,09,12,09,06,08,07,
12,09,07,07,06,06,08,06,15,08,06,06,10,10,10,07,05,10,07,11,
09,12,10,12,04,10,05,05,04,14,07,10,09,07,11,10,10,10,11,15,
09,14,12,09,09,07,12,04,10,10,06,10,07,28,06,10,08,09,10,10,
10,13,12,08,10,09,09,07,09,09,07,11,11,13,08,10,07
"""

df = get_list_of_university_towns().copy()

cols = ["State", "RegionName"]

print('Shape test: ', "Passed" if df.shape ==
      (517, 2) else 'Failed')
print('Index test: ',
      "Passed" if df.index.tolist() == list(range(517))
      else 'Failed')

print('Column test: ',
      "Passed" if df.columns.tolist() == cols else 'Failed')
print('\\n test: ',
      "Failed" if any(df[cols[0]].str.contains(
          '\n')) or any(df[cols[1]].str.contains('\n'))
      else 'Passed')
print('Trailing whitespace test:',
      "Failed" if any(df[cols[0]].str.contains(
          '\s+$')) or any(df[cols[1]].str.contains(
              '\s+$'))
      else 'Passed')
print('"(" test:',
      "Failed" if any(df[cols[0]].str.contains(
          '\(')) or any(df[cols[1]].str.contains(
              '\('))
      else 'Passed')
print('"[" test:',
      "Failed" if any(df[cols[0]].str.contains(
          '\[')) or any(df[cols[1]].str.contains(
              '\]'))
      else 'Passed')

states_vlist = [st.strip() for st in stateStr.split(',')]

mismatchedStates = df[~df['State'].isin(
    states_vlist)].loc[:, 'State'].unique()
print('State test: ', "Passed" if len(
    mismatchedStates) == 0 else "Failed")
if len(mismatchedStates) > 0:
    print()
    print('The following states failed the equality test:')
    print()
    print('\n'.join(mismatchedStates))

df['expected_length'] = [int(s.strip())
                         for s in regNmLenStr.split(',')
                         if s.strip().isdigit()]
regDiff = df[df['RegionName'].str.len() != df['expected_length']].loc[
    :, ['RegionName', 'expected_length']]
regDiff['actual_length'] = regDiff['RegionName'].str.len()
print('RegionName test: ', "Passed" if len(regDiff) ==
      0 else ' \nMismatching regionNames\n {}'.format(regDiff))


In [10]:
gdp = pd.read_excel('gdplev.xls')
gdp

Current-Dollar and "Real" Gross Domestic Product  \
0                                                NaN   
1                                             Annual   
2                                                NaN   
3                                                NaN   
4                                                NaN   
5                                                NaN   
6                                                NaN   
7                                               1929   
8                                               1930   
9                                               1931   
10                                              1932   
11                                              1933   
12                                              1934   
13                                              1935   
14                                              1936   
15                                              1937   
16                                              1938   
17                                              1939   
18                                              1940   
19                                              1941   
20                                              1942   
21                                              1943   
22                                              1944   
23                                              1945   
24                                              1946   
25                                              1947   
26                                              1948   
27                                              1949   
28                                              1950   
29                                              1951   
..                                               ...   
255                                              NaN   
256                                              NaN   
257                                              NaN   
258                                              NaN   
259                                              NaN   
260                                              NaN   
261                                              NaN   
262                                              NaN   
263                                              NaN   
264                                              NaN   
265                                              NaN   
266                                              NaN   
267                                              NaN   
268                                              NaN   
269                                              NaN   
270                                              NaN   
271                                              NaN   
272                                              NaN   
273                                              NaN   
274                                              NaN   
275                                              NaN   
276                                              NaN   
277                                              NaN   
278                                              NaN   
279                                              NaN   
280                                              NaN   
281                                              NaN   
282                                              NaN   
283                                              NaN   
284                                              NaN   

                             Unnamed: 1  \
0                                   NaN   
1                                   NaN   
2                                   NaN   
3                                   NaN   
4    GDP in billions of current dollars   
5                                   NaN   
6                                   NaN   
7                                 104.6   
8                                  92.2   
9                                  77.4   
10                                 59.5   
11             

In [61]:
def get_recession_start():
    '''Returns the year and quarter of the recession start time as a 
    string value in a format such as 2005q3'''
    
    # Clean GDP data frame
    gdp = pd.read_excel('gdplev.xls', skiprows=5)
    gdp = gdp.iloc[214:,4:6]
    gdp.columns = [['Quarter', 'GDP']]
    gdp = gdp.reset_index(drop=True)
    gdp['diff'] = gdp['GDP'].diff()
    gdp['diff2'] = [1 if x >= 0 else 0 for x in gdp['diff']]
    
    recession_start_quarters = [] #List of start quarters
    #Look up for the pattern [1]-0-0 to get recession [start]
    for i in range(len(gdp)-2):
        if ((gdp.loc[i, 'diff2'] == 1) & (gdp.loc[i+1, 'diff2'] == 0) &
            (gdp.loc[i+2, 'diff2'] == 0)):
            #Append to list if found:
            recession_start_quarters.append(gdp.loc[i, 'Quarter'])

    return recession_start_quarters[0] #Return just string due to requirements

get_recession_start()

'2008q3'

In [96]:
# Clean GDP data frame
gdp = pd.read_excel('gdplev.xls', skiprows=5)
gdp = gdp.iloc[214:,4:6]
gdp.columns = [['Quarter', 'GDP']]
gdp = gdp.reset_index(drop=True)
gdp['diff'] = gdp['GDP'].diff()
gdp['diff2'] = [1 if x >= 0 else 0 for x in gdp['diff']]
gdp[10:]

Quarter      GDP   diff  diff2
10  2002q3  11037.1  102.3      1
11  2002q4  11103.8   66.7      1
12  2003q1  11230.1  126.3      1
13  2003q2  11370.7  140.6      1
14  2003q3  11625.1  254.4      1
15  2003q4  11816.8  191.7      1
16  2004q1  11988.4  171.6      1
17  2004q2  12181.4  193.0      1
18  2004q3  12367.7  186.3      1
19  2004q4  12562.2  194.5      1
20  2005q1  12813.7  251.5      1
21  2005q2  12974.1  160.4      1
22  2005q3  13205.4  231.3      1
23  2005q4  13381.6  176.2      1
24  2006q1  13648.9  267.3      1
25  2006q2  13799.8  150.9      1
26  2006q3  13908.5  108.7      1
27  2006q4  14066.4  157.9      1
28  2007q1  14233.2  166.8      1
29  2007q2  14422.3  189.1      1
30  2007q3  14569.7  147.4      1
31  2007q4  14685.3  115.6      1
32  2008q1  14668.4  -16.9      0
33  2008q2  14813.0  144.6      1
34  2008q3  14843.0   30.0      1
35  2008q4  14549.9 -293.1      0
36  2009q1  14383.9 -166.0      0
37  2009q2  14340.4  -43.5      0
38  2009q3  14384.1   43.7      1
39  2009q4  14566.5  182.4      1
40  2010q1  14681.1  114.6      1
41  2010q2  14888.6  207.5      1
42  2010q3  15057.7  169.1      1
43  2010q4  15230.2  172.5      1
44  2011q1  15238.4    8.2      1
45  2011q2  15460.9  222.5      1
46  2011q3  15587.1  126.2      1
47  2011q4  15785.3  198.2      1
48  2012q1  15973.9  188.6      1
49  2012q2  16121.9  148.0      1
50  2012q3  16227.9  106.0      1
51  2012q4  16297.3   69.4      1
52  2013q1  16475.4  178.1      1
53  2013q2  16541.4   66.0      1
54  2013q3  16749.3  207.9      1
55  2013q4  16999.9  250.6      1
56  2014q1  17025.2   25.3      1
57  2014q2  17285.6  260.4      1
58  2014q3  17569.4  283.8      1
59  2014q4  17692.2  122.8      1
60  2015q1  17783.6   91.4      1
61  2015q2  17998.3  214.7      1
62  2015q3  18141.9  143.6      1
63  2015q4  18222.8   80.9      1
64  2016q1  18281.6   58.8      1
65  2016q2  18450.1  168.5      1

In [97]:
def get_recession_end():
    '''Returns the year and quarter of the recession end time as a 
    string value in a format such as 2005q3'''
    
    # Clean GDP data frame
    gdp = pd.read_excel('gdplev.xls', skiprows=5)
    gdp = gdp.iloc[214:,4:6]
    gdp.columns = [['Quarter', 'GDP']]
    gdp = gdp.reset_index(drop=True)
    gdp['diff'] = gdp['GDP'].diff()
    gdp['diff2'] = [1 if x >= 0 else 0 for x in gdp['diff']]
    
    recession_end_quarters = [] #List of start quarters
    #Look up for the pattern 0-0-1-[1] to get recession [end]
    for i in range(len(gdp)-3):
        if ((gdp.loc[i, 'diff2'] == 0) & (gdp.loc[i+1, 'diff2'] == 0) &
            (gdp.loc[i+2, 'diff2'] == 1) & (gdp.loc[i+3, 'diff2'] == 1)):
            #Append to list if found:
            recession_end_quarters.append(gdp.loc[i+3, 'Quarter'])

    return recession_end_quarters[0] #Return just string due to requirements

get_recession_end()

'2009q4'

In [101]:
def get_recession_bottom():
    '''Returns the year and quarter of the recession bottom time as a 
    string value in a format such as 2005q3'''
    
    # Clean GDP data frame
    gdp = pd.read_excel('gdplev.xls', skiprows=5)
    gdp = gdp.iloc[214:,4:6]
    gdp.columns = [['Quarter', 'GDP']]
    gdp = gdp.reset_index(drop=True)
    gdp['diff'] = gdp['GDP'].diff()
    gdp['diff2'] = [1 if x >= 0 else 0 for x in gdp['diff']]
    
    #Slice the GDP data frame to the recession period
    gdp = gdp[(gdp['Quarter'] >= get_recession_start()) & 
              (gdp['Quarter'] <= get_recession_end())]
    #Get row where GDP is min()
    bottom = gdp[gdp['GDP'] == gdp['GDP'].min()]
    
    return bottom.iloc[0,0]

get_recession_bottom()

'2009q2'

In [133]:
def convert_housing_data_to_quarters():
    '''Converts the housing data to quarters and returns it as mean 
    values in a dataframe. This dataframe should be a dataframe with
    columns for 2000q1 through 2016q3, and should have a multi-index
    in the shape of ["State","RegionName"].
    
    Note: Quarters are defined in the assignment description, they are
    not arbitrary three month periods.
    
    The resulting dataframe should have 67 columns, and 10,730 rows.
    '''
    house_df = pd.read_csv('City_Zhvi_AllHomes.csv')
    house_df.drop(['Metro','CountyName','RegionID','SizeRank'],axis=1,inplace=1)
    #Set multi-index first
    house_df = house_df.set_index(["State", "RegionName"])
    
    #Group columns by quarters and slice from 2000q1 to 2016q3
    #From https://stackoverflow.com/questions/40433717/pandas-how-to-use-groupby-to-group-columns-by-date-in-the-label
    house_df = house_df.groupby(pd.PeriodIndex(house_df.columns, freq='Q'), 
                                axis=1).mean().iloc[:,15:]
    return house_df

convert_housing_data_to_quarters()

2000Q1        2000Q2        2000Q3  \
State RegionName                                                      
NY    New York                      NaN           NaN           NaN   
CA    Los Angeles          2.070667e+05  2.144667e+05  2.209667e+05   
IL    Chicago              1.384000e+05  1.436333e+05  1.478667e+05   
PA    Philadelphia         5.300000e+04  5.363333e+04  5.413333e+04   
AZ    Phoenix              1.118333e+05  1.143667e+05  1.160000e+05   
NV    Las Vegas            1.326000e+05  1.343667e+05  1.354000e+05   
CA    San Diego            2.229000e+05  2.343667e+05  2.454333e+05   
TX    Dallas               8.446667e+04  8.386667e+04  8.486667e+04   
CA    San Jose             3.742667e+05  4.065667e+05  4.318667e+05   
FL    Jacksonville         8.860000e+04  8.970000e+04  9.170000e+04   
CA    San Francisco        4.305000e+05  4.644667e+05  4.835333e+05   
TX    Austin               1.429667e+05  1.452667e+05  1.494667e+05   
MI    Detroit              6.616667e+04  6.830000e+04  6.676667e+04   
OH    Columbus             9.436667e+04  9.583333e+04  9.713333e+04   
TN    Memphis              7.250000e+04  7.320000e+04  7.386667e+04   
NC    Charlotte            1.269333e+05  1.283667e+05  1.302000e+05   
TX    El Paso              7.626667e+04  7.686667e+04  7.673333e+04   
MA    Boston               2.069333e+05  2.191667e+05  2.331000e+05   
WA    Seattle              2.486000e+05  2.556000e+05  2.625333e+05   
MD    Baltimore            5.966667e+04  5.950000e+04  5.883333e+04   
CO    Denver               1.622333e+05  1.678333e+05  1.743333e+05   
DC    Washington           1.377667e+05  1.442000e+05  1.487000e+05   
TN    Nashville            1.138333e+05  1.152667e+05  1.158667e+05   
WI    Milwaukee            7.803333e+04  7.906667e+04  8.103333e+04   
AZ    Tucson               1.018333e+05  1.029667e+05  1.044667e+05   
OR    Portland             1.528000e+05  1.547667e+05  1.565667e+05   
OK    Oklahoma City        7.643333e+04  7.750000e+04  7.856667e+04   
NE    Omaha                1.128000e+05  1.141000e+05  1.167333e+05   
NM    Albuquerque          1.258667e+05  1.267000e+05  1.264333e+05   
CA    Fresno               9.410000e+04  9.526667e+04  9.646667e+04   
...                                 ...           ...           ...   
TX    Granite Shoals                NaN           NaN           NaN   
MD    Piney Point          1.556667e+05  1.551667e+05  1.584667e+05   
WI    Maribel                       NaN           NaN           NaN   
ID    Middleton            1.060667e+05  1.043333e+05  1.019000e+05   
CO    Bennett              1.329000e+05  1.358333e+05  1.398000e+05   
NH    East Hampstead       1.618333e+05  1.691000e+05  1.739667e+05   
MO    Garden City                   NaN           NaN           NaN   
AR    Mountainburg         5.716667e+04  6.433333e+04  6.783333e+04   
WI    Oostburg             1.072667e+05  1.081000e+05  1.124333e+05   
CA    Twin Peaks           9.736667e+04  1.001667e+05  1.013333e+05   
NY    Upper Brookville     1.230967e+06  1.230967e+06  1.237700e+06   
HI    Volcano              9.870000e+04  1.053667e+05  1.146667e+05   
SC    Wedgefield                    NaN           NaN           NaN   
MI    Williamston          1.591667e+05  1.613000e+05  1.643000e+05   
AR    Decatur              6.360000e+04  6.440000e+04  6.566667e+04   
TN    Briceville           4.000000e+04  4.173333e+04  4.366667e+04   
IN    Edgewood             9.170000e+04  9.186667e+04  9.293333e+04   
TN    Palmyra                       NaN           NaN           NaN   
MD    Saint Inigoes        1.480667e+05  1.476000e+05  1.572333e+05   
IN    Marysville                    NaN           NaN           NaN   
CA    Forest Falls         1.135333e+05  1.144000e+05  1.141667e+05   
MO    Bois D Arc           1.078000e+05  1.069667e+05  1.071000e+05   
VA    Henrico              1.285667e+05  1.307667e+05  1.322667e+05   
NJ    Diamond Beach        1.739667e+05  1.831000e+05  1.

In [17]:
def run_ttest():
    '''First creates new data showing the decline or growth of housing prices
    between the recession start and the recession bottom. Then runs a ttest
    comparing the university town values to the non-university towns values, 
    return whether the alternative hypothesis (that the two groups are the same)
    is true or not as well as the p-value of the confidence. 
    
    Return the tuple (different, p, better) where different=True if the t-test is
    True at a p<0.01 (we reject the null hypothesis), or different=False if 
    otherwise (we cannot reject the null hypothesis). The variable p should
    be equal to the exact p value returned from scipy.stats.ttest_ind(). The
    value for better should be either "university town" or "non-university town"
    depending on which has a lower mean price ratio (which is equivilent to a
    reduced market loss).'''
    
    return "ANSWER"